font/src/directwrite/mod.rs

```diff
@@ -13,33 +13,207 @@
+        // does not exist in the font
+        if glyph_index == 0 {
+            return Err(Error::MissingGlyph(glyph.c));
+        }
+
+        let em_size = em_size(glyph.size);
+
+        let glyph_run = dwrote::DWRITE_GLYPH_RUN {
+            fontFace: unsafe { font.as_ptr() },
+            fontEmSize: em_size,
+            glyphCount: 1,
+            glyphIndices: &(glyph_index),
+            glyphAdvances: &(0.0),
+            glyphOffsets: &(offset),
+            isSideways: 0,
+            bidiLevel: 0,
+        };
+
+        let rendering_mode = font.get_recommended_rendering_mode_default_params(
+            em_size,
+            self.device_pixel_ratio,
+            dwrote::DWRITE_MEASURING_MODE_NATURAL,
+        );
+
+        let glyph_analysis = GlyphRunAnalysis::create(
+            &glyph_run,
+            self.device_pixel_ratio,
+            None,
+            rendering_mode,
+            dwrote::DWRITE_MEASURING_MODE_NATURAL,
+            0.0,
+            0.0,
+        )
+        .or_else(|_| Err(Error::MissingGlyph(glyph.c)))?;
+
+        let bounds = glyph_analysis
+            .get_alpha_texture_bounds(dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1)
+            .or_else(|_| Err(Error::MissingGlyph(glyph.c)))?;
+
+        let buf = glyph_analysis
+            .create_alpha_texture(dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1, bounds)
+            .or_else(|_| Err(Error::MissingGlyph(glyph.c)))?;
+
+        Ok(RasterizedGlyph {
+            c: glyph.c,
+            width: (bounds.right - bounds.left) as i32,
+            height: (bounds.bottom - bounds.top) as i32,
+            top: -bounds.top,
+            left: bounds.left,
+            buf: BitmapBuffer::RGB(buf),
+        })
+    }
+
+    fn rasterize_fallback_glyph(
+        &mut self,
+        glyph: GlyphKey,
+        original_glyph: GlyphKey,
+    ) -> Result<RasterizedGlyph, Error> {
+        // The fallback font is unlikely to have the same advance size as the monospace font.
+        // Therefore we may need to dynamically decrease the font size so the glyph fits in the
+        // cell width.
+        //
+        // We'll only do this for any glyph > 2 cells wide, downsizing it to be just 2 cells wide.
+        // This should help keep things tidy without resizing many glyphs needlessly.
+
+        let font_metrics = self.metrics(original_glyph.font_key, original_glyph.size)?;
+        let glyph_metrics = self.metrics_from_glyph(glyph)?;
+
+        let max_advance = 2.0 * font_metrics.average_advance;
+        let actual_advance = glyph_metrics.average_advance;
+
+        if actual_advance > max_advance {
+            let scale_factor = (max_advance / actual_advance) as f32;
+
+            let mut scaled_glyph = glyph;
+            scaled_glyph.size = Size::new(glyph.size.as_f32_pts() * scale_factor);
+
+            self.rasterize_glyph(scaled_glyph)
+        } else {
+            self.rasterize_glyph(glyph)
+        }
```

Comment:
Reusing `rasterize_glyph` is nice, but calling it twice here still seems a bit strange to me.

Can't we just mutate `glyph` in the `if` block and then always call `self.rasterize_glyph(glyph)` afterwards?